File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int getHeight(TreeNode* root, bool& isBalanced){
        if (!root) return 0;
        int left = getHeight(root->left, isBalanced);
        int right = getHeight(root->right, isBalanced);

        if (isBalanced){
            isBalanced = abs(left-right) <= 1;
        } 
        return 1 + max(left, right);
    }
public:
    bool isBalanced(TreeNode* root) {
        bool isBalanced = true;
        getHeight(root, isBalanced);
        return isBalanced;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private void DFS(TreeNode node, int k, List<int> list){
        if (node == null) return;

        DFS(node.left, k, list);

        if (list.Count < k){
            list.Add(node.val);
        }
        else return;

        DFS(node.right, k, list);
    }
    public int KthSmallest(TreeNode root, int k) {
        var res = new List<int>();
        DFS(root, k, res);
        return res[^1];
    }
}
